File: src/org/genyris/io/JlineStdioInStream.java
package org.genyris.io;

import java.io.IOException;
import java.io.OutputStream;

import jline.console.ConsoleReader;
import jline.console.completer.ArgumentCompleter;
import jline.console.completer.ArgumentCompleter.AbstractArgumentDelimiter;
import jline.console.completer.StringsCompleter;

import org.apache.commons.io.output.WriterOutputStream;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;

public class JlineStdioInStream implements InStream {

    //
    // WARNING - this class has only one instance shared between all threads.
    //

    private int _nextIndex;
    private ConsoleReader _jline;
    private String _nextLine;
    private ArgumentCompleter _completer;
    private Environment _environment;
    private static JlineStdioInStream singleton = null;
    private static Interpreter _interp;
    private int _lineCount;

    public static synchronized JlineStdioInStream knew() { // the 'k' is silent.
        if (singleton == null) {
            singleton = new JlineStdioInStream();
        }
        return singleton;
    }

    private JlineStdioInStream() {
        _nextLine = null;
        _lineCount = 0;
        try {
            _jline = new ConsoleReader();
            _jline.setPrompt("> ");
            _jline.setExpandEvents(false);
            _completer = null;
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public OutputStream getOutput() {
        return new WriterOutputStream(_jline.getOutput());
    }

    public synchronized void unGet(char x) throws LexException {
        throw new LexException("StdioStream: unGet not implemented.");
    }

    public synchronized char readNext() throws LexException {
        if (_nextLine == null) {
            throw new LexException(
                    "StdioInStream: readNext() called before hasData()");
        }
        int retval = _nextLine.charAt(_nextIndex);
        _nextIndex++;
        if (_nextLine.length() == _nextIndex) {
            // read the last one
            _nextLine = null;
            _nextIndex = 0;
        }
        return (char) retval;
    }

    public synchronized boolean hasData() {
        try {
            if (_nextLine != null) {
                return true;
            }
            _nextLine = _jline.readLine();
            _lineCount++;
            _nextIndex = 0;
        } catch (IOException e) {
            _nextLine = null;
            return false;
        }
        if (_nextLine == null) {
            return false;
        } else {
            _nextLine = _nextLine + '\n';
            return true;
        }
    }

    private void refreshTabCompletion() {
        if (_environment == null) {
            _environment = _interp.getGlobalEnv();
        }
        _jline.removeCompleter(_completer);
        _completer = new ArgumentCompleter(new GenyrisArgumentDelimiter(),
                new StringsCompleter(
                        _interp.getBoundSymbolsAsListOfStrings(_environment)));
        _completer.setStrict(false);
        _jline.addCompleter(_completer);
    }

    public synchronized void close() throws GenyrisException {
    }

    public synchronized void resetAfterError() {
        _nextLine = null;
        beginningExpression();
    }

    public void setInterpreter(Interpreter _interpreter) {
        _interp = _interpreter;
        _environment = _interp.getGlobalEnv();
        refreshTabCompletion();
    }

    public void setEnvironment(Environment env) {
        _environment = env;
        refreshTabCompletion();
    }

    public void withinExpression(Environment env) {
        _environment = env;
        refreshTabCompletion();
        _jline.setPrompt(": ");
    }

    public void beginningExpression() {
        refreshTabCompletion();
        _jline.setPrompt("> ");
    }

    public static class GenyrisArgumentDelimiter extends
            AbstractArgumentDelimiter {

        public boolean isDelimiterChar(final CharSequence buffer, final int pos) {
            char ch = buffer.charAt(pos);
            return Character.isWhitespace(ch) || ch == '(' || ch == ')'
                    || ch == '[' || ch == ']' || ch == ','
                    || ch == '\'' || ch == '"' || ch == '#'
                    || ch == '^' || ch == '{' || ch == '}';
        }
    }

    public int getLineNumber() {
        return _lineCount;
    }
    public String getFilename() {
        return "console";
    }

}

File: src/org/genyris/io/JlineStdioInStream.kt
package org.genyris.io

import jline.console.ConsoleReader
import jline.console.completer.ArgumentCompleter
import jline.console.completer.ArgumentCompleter.AbstractArgumentDelimiter
import jline.console.completer.StringsCompleter
import org.apache.commons.io.output.WriterOutputStream
import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import org.genyris.interp.Interpreter
import java.io.IOException
import java.io.Reader

class JlineStdioInStream private constructor() : InStream {
    //
    // WARNING - this class has only one instance shared between all threads.
    //
    private var _nextIndex = 0
    private val _jline: ConsoleReader? = null
    private var _nextLine: String? = null
    private var _completer: ArgumentCompleter? = null
    private var _environment: Environment? = null
    private var _lineCount = 0

    @kotlin.Throws(UnsupportedOperationException::class)
    override fun getReader(): Reader? {
        throw UnsupportedOperationException()
    }

    init {
        try {
            _jline = ConsoleReader()
            _jline.setPrompt("> ")
            _jline.setExpandEvents(false)
            _completer = null
        } catch (e: IOException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }
    }

    val output: OutputStream
        get() = WriterOutputStream(_jline!!.getOutput())

    @kotlin.jvm.Synchronized
    @kotlin.Throws(LexException::class)
    override fun unGet(x: Char) {
        throw LexException("StdioStream: unGet not implemented.")
    }

    @kotlin.jvm.Synchronized
    @kotlin.Throws(LexException::class)
    override fun readNext(): Char {
        if (_nextLine == null) {
            throw LexException(
                "StdioInStream: readNext() called before hasData()"
            )
        }
        val retval: Int = _nextLine.charAt(_nextIndex).code
        _nextIndex++
        if (_nextLine.length() == _nextIndex) {
            // read the last one
            _nextLine = null
            _nextIndex = 0
        }
        return retval.toChar()
    }

    @kotlin.jvm.Synchronized
    override fun hasData(): Boolean {
        try {
            if (_nextLine != null) {
                return true
            }
            _nextLine = _jline!!.readLine()
            _lineCount++
            _nextIndex = 0
        } catch (e: IOException) {
            _nextLine = null
            return false
        }
        if (_nextLine == null) {
            return false
        } else {
            _nextLine = _nextLine + '\n'
            return true
        }
    }

    private fun refreshTabCompletion() {
        if (_environment == null) {
            _environment = _interp!!.getGlobalEnv()
        }
        _jline!!.removeCompleter(_completer)
        _completer = ArgumentCompleter(
            GenyrisArgumentDelimiter(),
            StringsCompleter(
                _interp!!.getBoundSymbolsAsListOfStrings(_environment)
            )
        )
        _completer!!.setStrict(false)
        _jline.addCompleter(_completer)
    }

    @kotlin.jvm.Synchronized
    @kotlin.Throws(GenyrisException::class)
    override fun close() {
    }

    @kotlin.jvm.Synchronized
    override fun resetAfterError() {
        _nextLine = null
        beginningExpression()
    }

    fun setInterpreter(_interpreter: Interpreter) {
        _interp = _interpreter
        _environment = _interp!!.getGlobalEnv()
        refreshTabCompletion()
    }

    fun setEnvironment(env: Environment?) {
        _environment = env
        refreshTabCompletion()
    }

    override fun withinExpression(env: Environment?) {
        _environment = env
        refreshTabCompletion()
        _jline!!.setPrompt(": ")
    }

    override fun beginningExpression() {
        refreshTabCompletion()
        _jline!!.setPrompt("> ")
    }

    class GenyrisArgumentDelimiter : AbstractArgumentDelimiter() {
        override fun isDelimiterChar(buffer: CharSequence, pos: Int): Boolean {
            val ch: Char = buffer.charAt(pos)
            return Character.isWhitespace(ch) || ch == '(' || ch == ')' || ch == '[' || ch == ']' || ch == ',' || ch == '\'' || ch == '"' || ch == '#' || ch == '^' || ch == '{' || ch == '}'
        }
    }

    override fun getLineNumber(): Int {
        return _lineCount
    }

    override fun getFilename(): String {
        return "console"
    }

    companion object {
        private var singleton: JlineStdioInStream? = null
        private var _interp: Interpreter? = null

        @kotlin.jvm.Synchronized
        fun knew(): JlineStdioInStream { // the 'k' is silent.
            if (singleton == null) {
                singleton = JlineStdioInStream()
            }
            return singleton!!
        }
    }
}
